### ISSUE TITLE
Algebraic Loops (fmiGetXXX/fmiSetXXX)

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/fmi-standard#138]
**Reported by akorobkine on 19 Jun 2013 13:40 UTC**
We have some concerns regarding the efficiency issues that may be introduced with clarification for handling of algebraic loops on page 66.

The description states that the algebraic loop can be solved using only fmiGetXXX and fmiSetXXX.
The possible problem here is that the input of an FMU block may not only affect the current state, but may impact upon non-differential variables, variables requiring projection, or variables that can result in a change in event variable states.

Prior to the discussion of the algebraic loop handling, this could be be performed in fmiCompletedIntegratorStep (with the event handling in fmiEventUpdate), but with the current design, handling of the algebraic loops (at every minor step) will trigger re-computation continuously. In addition, given that only fmiGetXXX and fmiSetXXX are needed for handling of these loops, any fmiGetXXX following a call to fmiSetXXX may require re-computation, as there is no way for the callback to know if it is in an algebraic loop.

We are thinking that we might need a discussion of re-computation of a model's states/variable, and possibly the addition of a new fmi callback, say fmiUpdateModel (or something similar) which will be used to force an update when needed, after the fmiSetXXX calls and before the fmiGetXXX calls that depend on the fmiSetXXX calls.

With this addition, only algebraic loops will see the efficiency hit, while the rest of the mainstream processing will not have to potentially induce a re-computation on a per-fmiGetXXX basis.

Thanks,
MapleSim / Maplesoft

----

Migrated-From: https://trac.fmi-standard.org/ticket/138

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 21 Jun 2013 12:03 UTC**
From your description, the issue and your solution is not clear to me. Maybe you can provide a small example that explains the problem more concretely. This is what I  understood:

A tool might not solve all algebraic systems all the time, e.g., during Continuous-Time Mode solve an index 2 DAE (that is solve only constraints on velocity level), and whenever the drift of the constraint is too large, project the solution on to the position level.

Since there is no essential change for the Continuous-Time Mode, it is still possible to perform the projection in fmiCompletedIntegratorStep, and there seems to be no difference to FMI 1.0 or FMI 2.0 Beta 4. So, this problem would have been present already in previous versions.

**Comment by modelica-trac-importer** [comment]
**Comment by akorobkine on 21 Jun 2013 16:53 UTC**
During the continuous integration, it is expected that constraint drift should be small as the variables in the model are changing continuously, so I can see that ignoring the constraints and event triggers during algebraic loops in that case may work.

During the initialization phase, however, the new initial values provided for the model can be significantly different from the original initial conditions the model was built for, and the application of the Newton iteration to stabilize the model might require projection and re-computation of the event trigger states.

For example, a pendulum model with spatial coordinates x(t), y(t) and a Lagrange multiplier F(t) would have position and velocity constraints given as:

x(t)<sup>2+y(t)</sup>2-1 = 0
2\*x(t)\*x'(t)+2\*y(t)\*y'(t) = 0

respectively.

This model could have been created with an event that triggers when y(t) passes through zero (i.e. when the mass passes through the horizontal plane passing through the anchor), and the model could have originally been created with the initial values:

(x,y,x',y') = (1/sqrt(2),1/sqrt(2),1/sqrt(2),-1/sqrt(2))

Now say that the model is to be re-initialized with y'(0)=-1/sqrt(2).

This requires a change in all the co-ordinate values, and a change in the state of the initial event trigger (from (y>0) = true to (y>0) = false), and during the process where this recomputation occurs constraints and event triggers cannot be ignored.

This suggests that the combined fmiGetXXX/fmiSetXXX calls in the algebraic loop handling need to support use of projection and event trigger checking durring the initialization phase.

I suppose this could be accomplished by running a different computation for updating the values after a fmiSetXXX when calling a fmiGetXXX when the FMU is in initialization mode. Is this the intent?

Thanks.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 24 Jun 2013 11:46 UTC**
Replying to [comment:2 akorobkine]:


> During the continuous integration, it is expected that constraint drift should be small as the variables in the model are changing continuously, so I can see that ignoring the constraints and event triggers during algebraic loops in that case may work.
> 
> During the initialization phase, however, the new initial values provided for the model can be significantly different from the original initial conditions the model was built for, and the application of the Newton iteration to stabilize the model might require projection and re-computation of the event trigger states.
> 
> For example, a pendulum model with spatial coordinates x(t), y(t) and a Lagrange multiplier F(t) would have position and velocity constraints given as:
> 
> x(t)<sup>2+y(t)</sup>2-1 = 0
> 2\*x(t)\*x'(t)+2\*y(t)\*y'(t) = 0
> 
> respectively.
> 
> This model could have been created with an event that triggers when y(t) passes through zero (i.e. when the mass passes through the horizontal plane passing through the anchor), and the model could have originally been created with the initial values:
> 
> (x,y,x',y') = (1/sqrt(2),1/sqrt(2),1/sqrt(2),-1/sqrt(2))
> 
> Now say that the model is to be re-initialized with y'(0)=-1/sqrt(2).
> 
> This requires a change in all the co-ordinate values, and a change in the state of the initial event trigger (from (y>0) = true to (y>0) = false), and during the process where this recomputation occurs constraints and event triggers cannot be ignored.
> 
> This suggests that the combined fmiGetXXX/fmiSetXXX calls in the algebraic loop handling need to support use of projection and event trigger checking durring the initialization phase.
> 
> I suppose this could be accomplished by running a different computation for updating the values after a fmiSetXXX when calling a fmiGetXXX when the FMU is in initialization mode. Is this the intent?
> 
> Thanks.

Yes, the idea is that fmiGetXXX usually performs a different computation in InitializationMode, as when in ContinuousTimeMode or EventMode. So one strategy can be to define that all outputs depend on all inputs in InitializationMode (which is the default). This means that the environment has to set all inputs with fmiSetXXX, before any fmiGetXXX can be called. The first fmiGetXXX that is called after fmiSetXXX in InitializationMode, computes all variables of the model, including the projection step. Subsequent fmiGetXXX calls just return the cached value from this first fmiGetXXX call.

However, there is one (uncritical) property that is new for me: In the scenario above, the user is able to set 4 start values, but there are constraints between these values. If inconsistent start values are set, the FMU corrects them with the projection. From an FMU point of view, these 4 start values are "guess" values (since the FMU is allowed to change them). According to the mathematical model, "guess" values must be set before calling fmiEnterModelInitializationMode and can no longer be set during initialization. This looks o.k.

To summarize, it seems that the type of FMUs that you want to handle, are supported by the proposal. If yes, please close this ticket.

**Comment by modelica-trac-importer** [comment]
**Comment by akorobkine on 25 Jun 2013 15:07 UTC**
Thanks for the explanation.  I am closing the ticket.

**Comment by modelica-trac-importer** [comment]
**Modified by akorobkine on 25 Jun 2013 15:07 UTC**
